posts/improving-the-access-control-of-a-jhipster-application.md

```diff
@@ -0,0 +1,222 @@
+
+# Backend customization
+
+## Spring Security and JHipster authorities
+
+JHipster uses [Spring Security](https://spring.io/projects/spring-security) to secure the application, it is a very mature and robust framework that can be used to secure Spring-based applications. Spring Security can be easily customized to change the application's authentication and access-control to fulfill any desired requirements. JHipster uses 4 kind of users (**system**, **anonymousUser**, **user** and **admin**) that have one or multiple authorities (**ROLE_ANONYMOUS**, **ROLE_USER** and **ROLE_ADMIN**). More information about the type of users and authorities can be found on the [JHipster security page](https://www.jhipster.tech/security/).
+
+User's access-control can be configured using authorities at two different levels: 
+- Url
+    - **user** can use the entities API
+    - only **admin** can use the management API
+- HTTP verb
+    - **user** can get a user using the verb `GET`
+    - only **admin** can remove a user using the verb `DELETE`
+
+Thereby, the application's API access-control can be configured at an url level or more specifically at a HTTP verb level. That gives us some nice flexibility and allows us to configure any access-control policy.
+
+## Customizing url with patterns
+
+The method `configure` in `SecurityConfiguration.java` configures the security and the part after `authorizeRequests()` is used to control the urls.
+```java
+.authorizeRequests()
+.antMatchers("/api/register").permitAll()
+.antMatchers("/api/activate").permitAll()
+.antMatchers("/api/authenticate").permitAll()
+.antMatchers("/api/account/reset-password/init").permitAll()
+.antMatchers("/api/account/reset-password/finish").permitAll()
+.antMatchers("/api/**").authenticated()
+.antMatchers("/management/health").permitAll()
+.antMatchers("/management/info").permitAll()
+.antMatchers("/management/**").hasAuthority(AuthoritiesConstants.ADMIN)
+```
+
+As seen above, the default behavior of a JHipster application chains all the rules using the `antMatchers` method. The pattern value `**` simply means that it will match any request so `"/api/**"` matches the application's API.
+
+More details about the methods used after the url matcher:
+- `permitAll` allows anyone
+- `authenticated` allows only authenticated users
+- `hasAuthority` allows only authenticated users with a given authority
+
+So for example, having `.antMatchers("/api/**").hasAuthority(AuthoritiesConstants.ADMIN)` will only allow **admin** to use the application's API.
+
+## Customizing an endpoint for a specific HTTP verb
+
+In some cases, the access-control will be based on the HTTP verb and it can't be done using the url patterns. In this case, the configuration must be done at the method level in the resource.
```

Comment:
```suggestion
In some cases, the access-control will be based on the HTTP verb and it can't be done using the URL patterns. In this case, the configuration must be done at the method level in the resource.
```